opal/mca/memory/patcher/memory_patcher_component.c

```diff
@@ -0,0 +1,432 @@
+        memmove(buffer, ptr, read_offset);
+    }
+
+ out_close:
+    close(fd);
+    return seg_size;
+}
+
+static int (*original_shmdt) (const void *);
+
+static int intercept_shmdt (const void *shmaddr)
+{
+    OPAL_PATCHER_BEGIN;
+    int result;
+
+    opal_mem_hooks_release_hook (shmaddr, memory_patcher_get_shm_seg_size (shmaddr), false);
+
+    if (original_shmdt) {
+        result = original_shmdt (shmaddr);
+    } else {
+        result = memory_patcher_syscall (SYS_shmdt, shmaddr);
+    }
+
+    OPAL_PATCHER_END;
+    return result;
+}
+#endif
+
+static int patcher_register (void)
+{
+    mca_memory_patcher_priority = 80;
+    mca_base_component_var_register (&mca_memory_patcher_component.super.memoryc_version,
+                                     "priority", "Priority of the patcher memory hook component",
+                                     MCA_BASE_VAR_TYPE_INT, NULL, 0, 0, OPAL_INFO_LVL_5,
+                                     MCA_BASE_VAR_SCOPE_CONSTANT, &mca_memory_patcher_priority);
+
+    return OPAL_SUCCESS;
+}
+
+static int patcher_query (int *priority)
+{
+    if (opal_patcher->patch_symbol) {
+        *priority = mca_memory_patcher_priority;
+    } else {
+        *priority = -1;
+    }
+    return OPAL_SUCCESS;
+}
+
+static int patcher_open (void)
+{
+    static int was_executed_already = 0;
+    int rc;
+
+    if (was_executed_already) {
+        return OPAL_SUCCESS;
+    }
+
+    was_executed_already = 1;
+
+    /* set memory hooks support level */
+    opal_mem_hooks_set_support (OPAL_MEMORY_FREE_SUPPORT | OPAL_MEMORY_MUNMAP_SUPPORT);
+
+#if 0
+    /* NTH: the only reason to hook mmap would be to detect memory protection. this does not invalidate
+     * any cache entries in the region. */
+    rc = opal_patcher->patch_symbol ("mmap", (uintptr_t) intercept_mmap, (uintptr_t *) &original_mmap);
+    if (OPAL_SUCCESS != rc) {
+        return rc;
+    }
+#endif
+
+    rc = opal_patcher->patch_symbol ("munmap", (uintptr_t)intercept_munmap, (uintptr_t *) &original_munmap);
+    if (OPAL_SUCCESS != rc) {
+        return rc;
+    }
+
+#if defined (SYS_mremap)
+    rc = opal_patcher->patch_symbol ("mremap",(uintptr_t)intercept_mremap, (uintptr_t *) &original_mremap);
+    if (OPAL_SUCCESS != rc) {
+        return rc;
+    }
+#endif
+
+#if defined (SYS_madvise)
+    rc = opal_patcher->patch_symbol ("madvise", (uintptr_t)intercept_madvise, (uintptr_t *) &original_madvise);
+    if (OPAL_SUCCESS != rc) {
+        return rc;
+    }
+#endif
+
+#if defined(SYS_shmdt) && defined(__linux__)
+    rc = opal_patcher->patch_symbol ("shmdt", (uintptr_t) intercept_shmdt, (uintptr_t *) &original_shmdt);
+    if (OPAL_SUCCESS != rc) {
+        return rc;
+    }
+#endif
+
+#if defined (SYS_brk)
+    rc = opal_patcher->patch_symbol ("brk", (uintptr_t)intercept_brk, (uintptr_t *) &original_brk);
+#endif
+
+    return rc;
+}
+
+static int patcher_close(void)
+{
+    /* NTH: is it possible to unpatch the symbols? */
```

Comment:
What happens when this component is unloaded -- do we get stale patched hooks?  I.e., if the user app calls `MPI_FINALIZE` and then calls `munmap()`, ... what happens?